File: Best Time to Buy and Sell Stock III/Best Time to Buy and Sell Stock III.java
public class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) return 0;
        int[] first = new int[prices.length];
        int[] second = new int[prices.length];
        
        int min = prices[0];
        for (int i = 1; i < prices.length; ++i) {
            min = Math.min(min, prices[i]);
            first[i] = Math.max(first[i - 1], prices[i] - min);
        }
        
        int max = prices[prices.length - 1];
        for (int i = prices.length - 2; i >=0; --i) {
            max = Math.max(max, prices[i]);
            second[i] = Math.max(second[i + 1], max - prices[i]);
        }
        
        int res = 0;
        for (int i = 0; i < prices.length; ++i) {
            res = Math.max(res, first[i] + second[i]);
        }
        return res;
    }
}

File: Best Time to Buy and Sell Stock III/Best Time to Buy and Sell Stock III.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        if (prices.size() == 0) return 0;
        int *first = new int[prices.size()];
        int *second = new int[prices.size()];
        
        int MIN = prices[0];
        first[0] = 0;
        for (int i = 1; i < prices.size(); ++i) {
            MIN = min(MIN, prices[i]);
            first[i] = max(first[i - 1], prices[i] - MIN);
        }
        
        int MAX = prices[prices.size() - 1];
        second[prices.size() - 1] = 0;
        for (int i = prices.size() - 2; i >=0; --i) {
            MAX = max(MAX, prices[i]);
            second[i] = max(second[i + 1], MAX - prices[i]);
        }
        
        int res = 0;
        for (int i = 0; i < prices.size(); ++i) {
            res = max(res, first[i] + second[i]);
        }
        return res;
    }
};